_posts/2021-01-09-Julia-1.6-what-has-changed-since-1.0.md

```diff
@@ -0,0 +1,629 @@
+There is a bit of a problem in that it isn't easy to make sure the colors don't get reused, as there are not many in the 16 color list (especially when you skip a few like black, white and error red).
+Due to how Pkg constructs its error messages, basically every package in the dependency graph gets some message prepared for it, but not displayed, so just assigning them a color in turn gets it to loop around which still results in colors being reused.
+There is a way to fix that, but it is a big change to add structured log messages that are colored only once when they are displayed.
+We decided after much debate to assign colors based on the hash of the package name and shortened UUID.
+This means things have consistent colors for any packages still listed, even after you have resolves some conflicts.
+I think this is going to be a subtle improvement on ease of use.
+As a cool hack, you can actually change the color list used.
+To get the much larger color list I wanted to use originally you can do:
+```julia
+append!(Pkg.Resolve.CONFLICT_COLORS,  [21:51; 55:119; 124:142; 160:184; 196:220])
+```
+
+
+The other change was consecutive lists of version numbers were compressed to continuous ranges.
+In 1.6, these ranges are split only if there is a version that exists between them that is not compatible.
+So normally we get just a single range, since compatibility is typically monotonic.
+In contrast, in 1.0, they were split if there was a potential version that could exist that is not compatible (even if that version doesn't currently exist).
+This means they split every time the rightmost nonzero part of the version number was incremented.
+For something with a lot of pre-1.0 versions, that is a lot of numbers.
+I think the new method is much cleaner and easier to read.
+
+
+## Time To First Plot
+People often complain about the "Time To First Plot" (TTFP) in Julia.
+I personally have never minded it -- by the time I am plotting something, I have done minutes of thinking so 10 seconds of compilation is nothing.
+Plotting, it turns out, is basically a really hard thing for a compiler.
+It is many, many, small methods, most of which are only called once.
+And unlike most Julia code, it doesn't actually benefit all that much from most of the work Julia's JIT is normally doing to specialize things -- plotting itself isn't in the hot-loop.
+To make a long-story short, plotting is the postcard example for Julia needing to compile things before it can run them.
+
+As a really hacky way to time this, consider the following timing.
+You can see I did `@time @time ...` and two numbers were displayed.
+The first number is how long from the start of the command until the `using Plots` and `plot(...)` command finished.
+The second is that, plus how long it took before it could show the output of the first.
+It's worth looking at this later because it turns out that invalidations (see below) made the REPL have to recompile a bunch of display code.
```

Comment:
Are you sure `@time @time` measures the time to plot?  After I see the second timing, I still need to wait a while to see the plot displayed.  I'm not sure `@time` keeps parsing after the `;`.  I think
```julia
@time (using Plots; display(plot(1:0.1:10, sin.(1:0.1:10))))
```
is more reliable, and `display` forces `@time` to take into account rendering the plot

Reply:
`@time @time (using Plots; display(plot(1:0.1:10, sin.(1:0.1:10))))` should be best.
The second `@time` is not to measure time to plot, but time until `print` works again.
Because in Julia 1.0 plotting invalidates `print` (or something like that)

Reply:
Maybe I shouldn't try and convey that in the blog post though.
Or at least not til i get to the invalidations section?